### PR TITLE
PFL Render – Post-Accessibility Testing

### PR DESCRIPTION
**Pull request by DesignCodex**
@jardakotesovec 

Hi Jarda,
I followed through with your recommendations and I managed to style the PFL as per the [updated design](https://www.figma.com/design/1krCkdv9MvIAx7w4ceVxqK/PF-Label?node-id=0-1), that John shared with me.

[d526f73](https://github.com/asmecher/pflPlugin/pull/12/commits/d526f73a433aec0837601ee1da8622188414dcd1) – this is a screenshot of the current render, in-browser:

Screenshot 2024-06-20 at 03 25 54

### Tables (3 columns + div roles)

I transformed the remaining tables into divs with the assigned roles.
I did exactly what you did on #11 , so it should be good to go.

I didn't understand why use the "Metric" string here (pfl-sr-only), but I replicated it anyway.

Screenshot 2024-06-20 at 03 05 05

### Needed to wrap divs around some elements for styling
In some instances, I created a div to wrap some elements in and be able to style them inline.

**This happened with an h3 and ul element:**

Screenshot 2024-06-20 at 03 15 24

**...and also with the dt and dd pairs**

Screenshot 2024-06-20 at 03 15 38

I hope this is not problematic accessibility-wise.

### Index List

I managed to add a description item in the $pflIndexList array (PflPlugin.inc.php) and so, each index entry has now a dynamic "aria-description" attribute. I also style these <a> elements with a min. of 24x24px.

Screenshot 2024-06-20 at 03 24 33

Screenshot 2024-06-20 at 03 24 48

### Closing thoughts
I also updated several translate keys overall and added them again, as well as some of the logic / conditional statements that were initially present.

Overall, I tried following your feedback as close as possible and I hope the small changes I made are not disruptive accessibility-wise.

Thank you very much!

### CODE REVIEW

**Comment by jardakotesovec** [comment]
@DesignCodex That all sounds good. Will dive into testing shortly.

- Divs don't have any semantical meaning - screen reader don't care about it, so you can use as many of them as necessary for styling purposes.
- The intention with pfl-sr-only "Metric" is that if you just listen to screen reader - than having empty first cell is not ideal. So telling that in this column is the "Metric" and in second its "This journal" and third "Other Journal" gives better idea what to expect. There might be better wording than "Metric" - but it needs to be translatable and this can be good initial value.

**Comment by jardakotesovec** [comment]
I did go through it via voice over and I like what I am hearing! I think its getting pretty good! :-)

@israelcefrin Would you mind to also take it for spin?

Only place where it got bit confusing was the Author statements table, because it does not have values in third column for other journals - voice over did not jump there at all.  But I expect we can assume to have always value there, right @asmecher ?

File: css/pfl.css [hunk 1]
```diff
@@ -1,364 +1,562 @@
...
-.publication-facts-label table th:last-of-type {
-  color: #848484;
-  font-weight: 600;
+.publication-facts-label p:dir(ltr) {
+  text-align: start;
+}
 
+.publication-facts-label p:dir(rtl) {
+  text-align: start;
 }
```
Comment (jardakotesovec) [inline css/pfl.css:379]: Very small thing. If its always text-align: start, maybe we could just remove the lrt/rtl variants here?

File: templates/pfl.tpl [hunk 2]
```diff
@@ -1,499 +1,236 @@
...
+
+            const ariaExpandedAttr = toggleButton.getAttribute("aria-expanded");
+
+            if (ariaExpandedAttr === "false") {
+                toggleButton.setAttribute("aria-expanded", "true")
+
+                pflContainer.classList.add("expanded");
+                highlightDropdown.classList.add("expanded");
+                placeHolderText.style.visibility = "hidden";
+
```
Comment (jardakotesovec) [inline templates/pfl.tpl:40]: This makes it basically empty button for screen reader - I think using pfl-sr-only class here would be better as that will still make it screen reader to read what button it is..   but woul be hidden for other users.

File: templates/pfl.tpl [hunk 3]
```diff
@@ -1,499 +1,236 @@
...
-            <span>{translate key="plugins.generic.pfl.close"}</span> <img class="pfl-info-icon" alt="" src="{$baseUrl}/plugins/generic/pflPlugin/img/close_icon.svg" />
-            </button>
-            <h1 id="pfl-modal-data-availability-title">{translate key="plugins.generic.pfl.modalTitle"}</h1>
-            <hr/>
-            {translate key="plugins.generic.pfl.information.dataAvailability" }
-            <p class="pfl-modal-footer">{translate key="plugins.generic.pfl.modalFooter"}</p>
-        </div>
-    </div>
+        <div id="pfl-fact-table" class="pfl-tables" role="region" tabindex="-1">
+
```
[the remark this answers is not in the document]
Comment (israelcefrin) [reply templates/pfl.tpl:73]: Hi @jardakotesovec, [the negative value in `tabindex `makes the element or component not reachable](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex#try_it) by tab navigation. If I understood correctly, this region has to be reachable, right? Hence, yes, we need to remove it.

An example that comes to mind  about removing an element from the tab navigation is a redundant link, an image and a title that goes to the same URL. It is just annoying hearing the same URL twice in a row because an image and a text link are next to each other on the DOM tree.

**Comment by asmecher** [comment]
@jardakotesovec, in response to [your question](https://github.com/asmecher/pflPlugin/pull/12#issuecomment-2180082315):

> Only place where it got bit confusing was the Author statements table, because it does not have values in third column for other journals - voice over did not jump there at all. But I expect we can assume to have always value there, right @asmecher ?

Referring to the Figma, I think there are supposed to be values in the 3rd column next to Data Availability and External Funding (under Author Statements), but they are missing from [the latest screenshot](https://github.com/asmecher/pflPlugin/pull/12#issue-2363385103). Those should be re-added. (Unless they were removed intentionally, @DesignCodex?)

**Comment by DesignCodex** [comment]
> @jardakotesovec, in response to [your question](https://github.com/asmecher/pflPlugin/pull/12#issuecomment-2180082315):
> 
> > Only place where it got bit confusing was the Author statements table, because it does not have values in third column for other journals - voice over did not jump there at all. But I expect we can assume to have always value there, right @asmecher ?
> 
> Referring to the Figma, I think there are supposed to be values in the 3rd column next to Data Availability and External Funding (under Author Statements), but they are missing from [the latest screenshot](https://github.com/asmecher/pflPlugin/pull/12#issue-2363385103). Those should be re-added. (Unless they were removed intentionally, @DesignCodex?)

Hi Alec,
Referring back to the original PFL plugin, these values are dynamic and vary depending on each article / journal (Yes / No).
Per my understanding, these values would only appear if, for example, "Data Availability" returns a "Yes".

Since in this working model it returns a "No" by default, they don't appear. This logic was built via a series of conditional statements throughout the different rows; here I'll show you a snippet:

Screenshot 2024-06-20 at 21 20 44

**Comment by DesignCodex** [comment]
Thanks for the inputs @jardakotesovec, @israelcefrin and @asmecher,

I successfully integrated all feedback on my latest commit [1108277](https://github.com/asmecher/pflPlugin/pull/12/commits/1108277c561d2b6df98ef78fba2a0ed1de011cef)

**Comment by asmecher** [comment]
@DesignCodex, re: [your comment](https://github.com/asmecher/pflPlugin/pull/12#issuecomment-2181500952): 

The numbers to the right of the "Yes" / "No" for "Data Availability" and "External Funding" should *always* appear, regardless of whether the statement is available for this particular article. The "Yes" / "No" are for this specific article, and the number provides a point of comparison for how often these statements are available in other venues.

**Comment by DesignCodex** [comment]
> @DesignCodex, re: [your comment](https://github.com/asmecher/pflPlugin/pull/12#issuecomment-2181500952):
> 
> The numbers to the right of the "Yes" / "No" for "Data Availability" and "External Funding" should _always_ appear, regardless of whether the statement is available for this particular article. The "Yes" / "No" are for this specific article, and the number provides a point of comparison for how often these statements are available in other venues.

@jardakotesovec, @israelcefrin and @asmecher,

I pushed another commit [24d84a6](https://github.com/asmecher/pflPlugin/pull/12/commits/24d84a618f7470173164a02f7aa1eb198934e5ca) addressing @asmecher's last comment.

Also included wider spacing between the title and the thick black bar on top (as @jardakotesovec suggested, in alignment with John).

Kind regards,

**Comment by israelcefrin** [comment]
Hi all,

The results seem pretty good using VoiceOVer. We will know when someone who relies only on SR uses it. :-)

I have a few comments on issues during my spin.

The ORCID image logo is just a decorative image, it is recommended to remove it from the navigation by tab.
To remove it, add `aria-hidden="true"` and `"tabindex="-1"`. It will be still visual, but SR will not announce` "orcid logo"`
image

Why do we need to keep "N/A" linked? Can we remove it? It is focusable using keyboard navigation and SR announces it as a link. It might cause some confusion since when you active it, the page refreshes.
image


Ok, now a final request. Given that we are providing a workaround to provide a look & feel of a nutrition facts label, I think that we still need to make available a CSV or tabular version of data with less visual information.  Therefore, is it possible to provide a link after the label to download the data as a tabular format, e.g.: CSV ? For some users, it is easier to read tabular data on a tabular way, and we already will have all the data organized before rendering the label.

**Comment by DesignCodex** [comment]
> Hi all,
> 
> The results seem pretty good using VoiceOVer. We will know when someone who relies only on SR uses it. :-)
> 
> I have a few comments on issues during my spin.
> 
> The ORCID image logo is just a decorative image, it is recommended to remove it from the navigation by tab. To remove it, add `aria-hidden="true"` and `"tabindex="-1"`. It will be still visual, but SR will not announce` "orcid logo"` image
> 
> Why do we need to keep "N/A" linked? Can we remove it? It is focusable using keyboard navigation and SR announces it as a link. It might cause some confusion since when you active it, the page refreshes. image
> 
> Ok, now a final request. Given that we are providing a workaround to provide a look & feel of a nutrition facts label, I think that we still need to make available a CSV or tabular version of data with less visual information. Therefore, is it possible to provide a link after the label to download the data as a tabular format, e.g.: CSV ? For some users, it is easier to read tabular data on a tabular way, and we already will have all the data organized before rendering the label.

Hi @israelcefrin,

### ORCIDs  tags
Already took care of the orcid's as you described, thanks!


### "N/A"
Regarding the "N/A" – that's just a placeholder I added for showcasing the table with text, as it should have the Academic Society title / name. This is a new row that was added in this version of the PFL and I didn't have the means to dynamically retrieve the info.

This was briefly discussed in one of Mattermost's threads but I don't know if the link was created.


### Link to CSV / Tabular info.
I think this can be easily included but since it will impact design, I suggest aligning this with John first.

I'll push an updated version once I get feedback on these two last points.
Kind regards,

**Comment by jardakotesovec** [comment]
> Link to CSV / Tabular info.

Not sure what to do about this one. Reason why we have this complicated html coding is because its only partially table data. So I don't see how this would be expressed in csv. If its straightforward express as csv we would code the whole thing as table in the first place.

File: css/pfl.css [hunk 4]
```diff
@@ -454,15 +384,14 @@
 /* ACCESSIBILITY */
 
 .pfl-sr-only {
-  position: absolute;
-  width: 1px;
-  height: 1px;
+  width: 1px !important;
+  height: 1px !important;
```
Comment (jardakotesovec) [inline css/pfl.css:388]: why the changes for pfl-sr-only? I took the sr-only definition from tailwindcss, which I would think is well tested css for "screen reader only" content..

File: templates/info.tpl [hunk 5]
```diff
@@ -12,6 +12,10 @@
 
 <link rel="stylesheet" href="{$baseUrl}/plugins/generic/pflPlugin/css/pfl.css" type="text/css">
 
+<style>
+@import url('https://fonts.googleapis.com/css2?family=Noto+Sans:ital,wdth,wght@0,62.5..100,100..900;1,62.5..100,100..900&display=swap');
+</style>
+
```
Comment (jardakotesovec) [inline templates/info.tpl:18]: @asmecher I believe we aim to self-host everything to avoid any kind of tracking, right?

File: templates/pfl.tpl [hunk 6]
```diff
@@ -63,13 +63,13 @@
  
     <div class="pfl-dropdown">
         <button id="pfl-button-open-facts" aria-controls="pfl-fact-table" aria-expanded="false">
-            <span id="pfl-buttonText">{translate key="plugins.generic.pfl.publicationFacts"}</span>
+            <span id="pfl-buttonText">{translate key="plugins.generic.pfl.publicationFacts"}</span><img src="{$baseUrl|concat:"/plugins/generic/pflPlugin/img/pfl-down-arrow.svg"}" alt="dropdown arrow button">
         </button>
```
Comment (jardakotesovec) [inline templates/pfl.tpl:67]: This is also just icon, which not relevant for SR - can be addressed by adding aria-hidden=true. So you can also remove the alt, as alt would be used for image which is actual content - than its important for screen reader to have some description to pick up.

**Comment by israelcefrin** [comment]
Hello, just adding a note related to the button that shows/hides the content. It would be great if we could provide a visual feedback of its interactive behaviour.
W3C recommends:

> To help people with visual impairments identify the disclosure as interactive and make it easier to perceive that clicking either the disclosure button or its label changes the expanded state, when a pointer hovers over the button or its label, the background color changes, a border appears, and the cursor changes to a pointer.

https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-image-description/#accessibilityfeatures

Also, this example shows how the label of a show/hide button has a persistently displayed label. Currently, we are hiding the button label when the "facts label" is shown. Furthermore, the icon "drop-down arrow" is an image using "alt" which is announced every time what is a bit annoying when you use SR and it doesn't add any relevant information to the component, e.g.: instructions, how to open/close it, etc...

Could we remove the icon from the navigation DOM (aria-hidden)? Otherwise, if we keep it we will need to prepare it to be translated since SRs announces it.

File: templates/info.tpl (hunk 5 above)
Comment (jardakotesovec) [reply templates/info.tpl:18]: @DesignCodex We download&provide all fonts from themes/plugins to be indepdendent from any 3rd party service.

**Comment by DesignCodex** [comment]
Hello everyone,
@jardakotesovec @israelcefrin @asmecher

### Pushed another commit: [aa0dee1](https://github.com/asmecher/pflPlugin/pull/12/commits/aa0dee1aaf92866ff06e42657a7f1f9055af4044)
- Hid ORCIDs + "down-arrow" svg from nav. ("aria-hidden=true" and "tabindex-1", didn't do it for the info. icon at the footer)
- Created an hover effect for the dropdown button
- Increased area of button for it to occupy the entire box (it had a padding that made the borders unclickable)
- Included the "Noto Sans" typeface (woff2) in the respective folder 
- Plus minor styling adjustments

@israelcefrin just to reinforce that now the button label doesn't "hide" anymore, we solved it via styling.
@jardakotesovec also managed to make the .pfl-sr-only class get back to default again.

Thank you and kind regards,
